### ISSUE TITLE
Gitea support

### ISSUE THREAD

**Issue by hulto** [cloudbase/garm#323]
Hey!
I just started playing with Garm and I'm really enjoying it!
I'd like to use it with my gitea server as well as some GitHub repos.

Are there plans to add support for Gitea? If not is that something I can contribute?

**Comment by gabriel-samfira** [comment]
Hi @hulto!

As long as Gitea/Forgejo supports:
* ephemeral runners
* registration tokens/JIT runners
* can send webhooks with job runs so we can schedule runners to a pool
* Has a way to allow us to list compatible agent downloads for linux/windows/mac/etc

It should be possible. However, I think there is an inherent danger in trying to support 2 different forges in the same project, as they may deviate significantly in the future and it may become cumbersome or downright impossible to consolidate. It may be wise to attempt to rearchitect parts of GARM and split off the bits that deal with the interaction between GARM and the forge. Similar to how providers work. But that would be a huge undertaking.

In the short term, if gitea has the above mentioned features, an attempt could be made to support it, if you have the time. Up to this point the focus has been to support GitHub and GHES. So the code mostly revolves around that assumption. But I think some thin abstractions could make gitea fit.

My experience with gitea is limited, so I am not sure how much I could help, but if I can, I will. If you want to tackle this, I suggest creating a proof of concept, hacked with an axe, just to see if it's feasible. Something that won't take months of your time, just to find that we can't merge it.

If having the two forges coexist does not seem possible, another option would be to fork GARM, yank out github and replace it with gitea. The we can merge back if/when the architecture of GARM permits pluggable forges, as we could then separately maintain plugins for gitea, github, etc.

So, plenty of options 😊.

**Comment by gabriel-samfira** [comment]
Hi @ChristopherHX 

Adding Gitea will be a bit more involved. There are a few assumptions in GARM that are related to GitHub. Each entity (Repo, Org, Enterprise) has a "pool manager" (needs to be reworked). That pool manager makes sure that the runners in each pool are properly created, deleted, etc. To ensure that GARM itselt has valid credentials to manage resources inside an entity, we periodically fetch the tools. We need them anyway, and in some cases, tools have a bearer token we need to ass to be able to download them (GHES). So checking for tools has a dual purpose:

* Validate credentials
* keep the tools object in cache for the entity

For Gitea we don't have enterprises, which is fine. But if Gitea doesn't have "organizations", only repos and global, we need to create a new entity type.

We also need to determine how webhooks work in Gitea. For example, in GitHub we can have an enterprise which has multiple orgs which in turn have multiple repos. I can install GARM at all 3 levels. If we have repo "testrepo" in org "testorg" which is part of "testenterprise" and I have GARM installed on all 3 levels, then whenever "testrepo" triggers a workflow, all 3 will send a webhook to GARM for the same job.

Does Gitea have a similar behavior? If so, we need a way to determine who triggered that webhook, so we know which pools are entitled to react and create a runner.

To get this to work we will need to:

* Abstract the forge specific code that deals with webhooks
* After determining the forge type, we need to:
  * Determine which entity types are available for this implementation
  * Save the job in the db and associate it with the entity
* Create a new `credentials` type for the new forge type. Github has apps and PATs. Something similar needs to exist in Gitea.
* Determine if JIT runners is officially supported, if not, this bit will be hidden away in the `pool manager` implementaion. 
* Abstract away the pool manager to consume jobs and create runners associated with a specific pool
  * Based on the origin forge, we need to generate proper userdata that sets up the runner and registers it in the forge

There are probably more points to consider, but if we are to add Gitea support, I need to cleanup the code a bit and remove some assumptions that are tightly coupled to GitHub. I would like to do this anyway, because I want us to be able to support Gitea as well as Forgejo. Porentially others in the future.

It might take a while though. In the meantime if you'd like to create a PoC, do not let me discourage you 🙂 .

**Comment by gabriel-samfira** [comment]
Hi @ChristopherHX 

Would you recommend we use go-github as the client to interact with Gitea or does Gitea maintain its own client? If we are to add support for Gitea, I would rather we use a proper client that is maintained by Gitea which may mirror any new additions to the API, etc.

I would like to take the proper time to do it right and make sure it is maintainable moving forward.

**Comment by gabriel-samfira** [comment]
The gitea API seems to have swagger annotations. I suspect that a client could be generated for all endpoints. GARM has something similar. Our client is completely generated.

> Sure, good idea. Keep in mind Forgejo (since you mentioned it in this issue) state is unknown about this, they seem to go another path my last state has been ephemeral runner protocol change rejected.

I am following their discussion as well. They seem to lack the ephemeral flag for their runners. The `one-job` command just tries to fetch one job, run it then exit. That is not really secure, unless the job is run in a container/VM which would never gain access to the machine that has the runner token set.

I would love to support both Gitea and Forgejo long term, but right now, gitea seems to have everything we expect for ephemeral runners. So might as well start there.

Webhooks are great to start with. What would be awesome (and this is just spit balling ony part) long term is to have some longpoll or websocket/grpc endpoint where we can subscribe for jobs. Something like how scale sets work in github. This would be good for multiple reasons:

* No need to expose the GARM api to the forge for webhooks. GARM would subscribe as a client.
* Webhooks may be triggered while GARM is down, in which case GARM will not know about pending jobs. This can be mitigated by idle runners, but if you're using a public cloud, this may incur cost.
* Efficient scheduling of runners. We can acquire a specific job by a specific entity (user, global, org) and spin up a runner for it. If we can't, we release the job and let other potential entities/auto scalers try. Or we can retry.

But for now, webhooks are great.

**Comment by gabriel-samfira** [comment]
> I now tried the third party k8s provider as well https://gitea.com/gitea/helm-actions/issues/8. This operator does not seem to need any extra spec, because it drops the runner setup template. I have built my own docker image using the expected abi and it works.
> 
> garm works great for my experiments 👍, the runner scaled from zero to n for a Gitea Actions matrix using the workflow_job metric.
> 
> I guess my idea of using the garm-operator with my custom default branch build didn't work correctly, because of some incompatible changes between release garm v1.5.0 and patched main or I just didn't understand how this thing works.

Amazing stuff!

I have a few days of time coming up soon (in a couple of weeks) and I intend to spend a few days to try and implement the forge abstraction thing so we can cleanly add Gitea.

**Comment by gabriel-samfira** [comment]
Apropos, there is a configurable delay between when a job is received and when GARM can react. By default it's 30 seconds. The rationale was to allow any existing idle runner pick up the job before we create a new one.

You can tweak it using:

```bash
garm-cli controller update --minimum-job-age-backoff 0
```

You can see the current backoff setting using:

```bash
garm-cli controller show
```

**Comment by gabriel-samfira** [comment]
Yup. With the k8s provider, it should be quite fast. LXD can also be optimized to start in seconds.

**Comment by gabriel-samfira** [comment]
Yup. When we split out the forge related code, we will most likely define an interface that will deal with webhooks/scalesets on github and webhooks on gitea. This interface will consume events and create jobs. For each implementation we will look in the appropriate place for the needed info.

We will then create a job with all required info. We will most likely either extend the current job model or create a new one for each forge type.

So I would like to do this cleanly and avoid shoehorning gitea into GitHub assumptions. It would be nice to have gitea as a first class citizen.

Shoehorning is fine for a PoC, though! Gives us an idea about the amount of work required, and the progress you made is absolutely amazing.

**Comment by gabriel-samfira** [comment]
That's interesting! I would still create that abstraction. Not because gitea is not compatible enough with github, but because if the need ever arizes for them to break that compatibility, we won't have to refactor the entire project to accommodate that change.

Gitea is it's own thing and I think we should show it the consideration it deserves. It also opens the way for other forges in the future, if we do it right for gitea.

I am hoping to start having a look by the end of next week.

**Comment by gabriel-samfira** [comment]
I've started generalizing the code a bit to make room for other forges. The goal is to be able to define endpoint types for github, gitea and whatever else we may add, potentially defining new tables to hold that info. To that end, the params are being generalized a bit and working my way through the code as that happens.

It is quite the rabbit hole and it might take a while as time has become quite limited.

**Comment by gabriel-samfira** [comment]
> For global and user runner support might be no other option than using a gitea sdk in long term.

Yeah. Initial support will probably have only Repos and Orgs. But we can iterate later.

**Comment by gabriel-samfira** [comment]
Making notes here; will probably need to report upstream after I finish:

* nightly version of gitea returns status code `500` if I try to remove a runner that does not exist
* once an ephemeral runner finishes it shuts down on the VM (expected) but lingers in gitea until I try to manually run it again. It errors out, but then gitea realizes it isn't supposed to be there and removes it.
* The `garm-cli repo webhook install` command doesn't work using the go-github client. Gitea requires webhook type. Will probably need to use the official SDK for that.

**Comment by gabriel-samfira** [comment]
> once an ephemeral runner finishes it shuts down on the VM (expected) but lingers in gitea until I try to manually run it again. It errors out, but then gitea realizes it isn't supposed to be there and removes it.

ahh. I also need the `--once` flag when starting the runner. Feels like if the `--ephemeral` flag is used during registration, it should not need the `--once` flag to ensure gitea removes the runner once it is done.

**Comment by gabriel-samfira** [comment]
> Hmm, I do not understand how this could be the case. The ephemeral flag is saved in `.runner` and auto applies --once.

have not looked too much into it. But if I also specify the `--once` flag, the runner doesn't linger after it runs a job. It gets removed from gitea. will double back and check after I finish.

**Comment by gabriel-samfira** [comment]
And we have a basic working branch (for repos at least):

https://github.com/user-attachments/assets/ca963028-576d-4b13-84ff-e1cb0588293a

**Comment by gabriel-samfira** [comment]
The branch lack a bunch of stuff, needs to be cleaned up and needs tests, but for anyone bored and brave enough, it can be tested.

I like the idea of using the same CLI to manage both github and gitea runners 😄

**Comment by gabriel-samfira** [comment]
I added some instructions on how to test in #393. For anyone new to GARM, the existing documentation for GitHub applies to Gitea as well. The instructions in the PR assumes that the existing docs have been read.

**Comment by gabriel-samfira** [comment]
There is now a "nightly" docker image that includes the latest `main` (including cli). You can find it here:

https://github.com/cloudbase/garm/pkgs/container/garm/420228094?tag=nightly

Feel free to try it out. I'll add a docker xompose example later today.

**Comment by techknowlogick** [comment]
Amazing, thanks for everyone's work on this :)

**Comment by gabriel-samfira** [comment]
> Amazing, thanks for everyone's work on this :)

It was a genuine pleasure. If you plan to [test it out](https://github.com/cloudbase/garm/blob/main/doc/gitea.md), any feedback is welcome.

**Comment by hulto** [comment]
This is sick! Trying it out right now 🙂

**Comment by techknowlogick** [comment]
@Femumme yes, most actions are built using node, and so unless you want to re-implement pre-built actions it is suggested that you have an image with node pre-installed.
